airframe-sql/src/main/scala/wvlet/airframe/sql/analyzer/TypeResolver.scala

```diff
@@ -295,43 +299,67 @@ object TypeResolver extends LogSupport {
   private def findMatchInInputAttributes(
       context: AnalyzerContext,
       expr: Expression,
-      inputAttributes: Seq[Attribute]
+      inputAttributes: Seq[Attribute],
+      isSelectItem: Boolean
   ): List[Expression] = {
     val resolvedAttributes = inputAttributes.map(resolveAttribute)
 
+    def matchedColumn(matched: Seq[Expression], name: String): Expression = {
+      matched match {
+        case attrs if attrs.size == 1 => attrs.head
+        case attrs                    => MultiColumn(attrs, Some(name), None)
+      }
+    }
+
     def lookup(name: String): List[Expression] = {
       QName(name, None) match {
         case QName(Seq(db, t1, c1), _) if context.database == db =>
           resolvedAttributes.collect {
             case a: ResolvedAttribute if a.matchesWith(t1, c1) => a.withQualifier(t1)
-            case c: SingleColumn if c.matchesWith(t1, c1)      => c.expr
+            case c: SingleColumn if c.matchesWith(t1, c1) =>
+              c.expr match {
+                case m: MultiColumn => matchedColumn(m.matched(t1, c1), s"${t1}.${c1}")
+                case other          => other
+              }
+            case c: AllColumns if c.matchesWith(t1, c1) => matchedColumn(c.matched(t1, c1), s"${t1}.${c1}")
           }.toList
         case QName(Seq(t1, c1), _) =>
           resolvedAttributes.collect {
             case a: ResolvedAttribute if a.matchesWith(t1, c1) => a.withQualifier(t1)
-            case c: SingleColumn if c.matchesWith(t1, c1)      => c.expr
+            case c: SingleColumn if c.matchesWith(t1, c1) =>
+              c.expr match {
+                case m: MultiColumn => matchedColumn(m.matched(t1, c1), s"${t1}.${c1}")
+                case other          => other
+              }
+            case c: AllColumns if c.matchesWith(t1, c1) => matchedColumn(c.matched(t1, c1), s"${t1}.${c1}")
           }.toList
         case QName(Seq(c1), _) =>
           resolvedAttributes.collect {
-            case a: ResolvedAttribute if a.name == c1 => a
-            case c: SingleColumn if c.matchesWith(c1) => c.expr
+            case a: ResolvedAttribute if a.matchesWith(c1) => a
+            case c: SingleColumn if c.matchesWith(c1) =>
+              c.expr match {
+                case m: MultiColumn => matchedColumn(m.matched(c1), c1)
+                case other          => other
+              }
+            case c: AllColumns if c.matchesWith(c1) => matchedColumn(c.matched(c1), c1)
           }.toList
         case _ =>
           List.empty
       }
-    }
+    }.distinct
 
     val results = expr match {
       case i: Identifier =>
-        lookup(i.value)
+        lookup(i.value).map {
+          case a: Attribute   => a
+          case m: MultiColumn => m
+          // retain alias for select column
+          case expr => if (isSelectItem) SingleColumn(expr, Some(i.value), None, expr.nodeLocation) else expr
```

Comment:
Retain column alias for expressions in SELECT clause.
- `select 1 as cnt` -> retain `cnt` as alias
- `select id, count(*) cnt from A group by id having cnt > 10` -> `having count(*) > 10` (No need to retain `cnt` in HAVING clause)